### PR TITLE
Enable WSGI for Python Function App

### PR DESCRIPTION
**Pull request by Hazhzeng**
### Acknowledgement
This feature is migrated from @vtbassmatt 's **[azf-wsgi](https://github.com/vtbassmatt/azf-wsgi)** project which enables wsgi feature for Azure Functions Python app. We've contacted @vtbassmatt and agreed to implement this feature on azure-functions-python-library. We greatly appreciate his contribution, helping us expand the Python user base in his own time.

### Usage
1. Override the main entrypoint
```python
import logging

import azure.functions as func
from ..FlaskApp.wsgi import application

main = func.WsgiMiddleware(application).main
```

2. Alternatively, use the old syntax as demonstrated in **[azf-wsgi](https://github.com/vtbassmatt/azf-wsgi)**
```python
import logging

import azure.functions as func
from ..FlaskApp.wsgi import application

def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
    return func.WsgiMiddleware(application).handle(req, context)
```

### Fixes
This PR also includes some fixes from the original repository
Thanks
- @caffeinate for creating the body_encoding fix https://github.com/vtbassmatt/azf-wsgi/pull/6
- @james1293 for creating the BytesIO fix https://github.com/vtbassmatt/azf-wsgi/pull/5

### Limitation
- The azure functions architecture does not support data streaming since it requires http persistent connection.
- The Python function app is running on a single process, to spawn multiple processes from the same host, set FUNCTIONS_WORKER_PROCESS_COUNT app setting to a higher value (10).

### Todos
Since the feature is not officially released, **feature-wise this PR is ready for review**, and of course, we need to verify the following features before we release this WSGI change.
1. Client app logging ✔
2. Client app routing (query_params) ✔
3. Environment variable propagating (from host to client app) ✔
4. Content-Length (binary, application/json, text/plain) ✔
5. Chunked data transfer ✔
6. Response headers ✔
7. Client app error emissions ✔
8. Client app exception handling ✔

resolves: https://github.com/Azure/azure-functions-python-worker/issues/165

### CODE REVIEW

**Comment by rolexsanches** [comment]
Hello

Its aready able to run Django Applications too?

Thanks!

**Comment by Hazhzeng** [comment]
Hey @rolexsanches,

Yes, Django also accept wsgi request via WSGI interface.
The following code should do the trick.
```python
import logging

import azure.functions as func
from ..Django.wsgi import DjangoApplication

main = func.WsgiMiddleware(DjangoApplication).main
```
 
Please wait for the next Python release to enable this feature.

File: azure/functions/_http_wsgi.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Implement interfaces for PEP 3333 environ
+        self.request_method = getattr(func_req, 'method', None)
+        self.script_name = ''
+        self.path_info = getattr(url, 'path', None)
+        self.query_string = getattr(url, 'query', None)
+        self.content_type = self._lowercased_headers.get('content-type')
+        self.content_length = str(len(func_req_body))
+        self.server_name = getattr(url, 'hostname', None)
+        self.server_port = str(self._get_port(url, self._lowercased_headers))
```
Comment (ankitkumarr) [inline azure/functions/_http_wsgi.py]: `getattr` might be slightly slower (due to reflection) than regular get, but may not be a big deal. Just a note.

File: azure/functions/_http_wsgi.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            return int(lowercased_headers['x-forwarded-port'])
+        elif getattr(parsed_url, 'port', None):
+            return parsed_url.port
+        elif parsed_url.scheme == 'https':
+            return 443
+        return port
+
+    def _get_http_headers(self,
+                          func_headers: Dict[str, str]) -> Dict[str, str]:
+        # Content-Type -> HTTP_CONTENT_TYPE
```
Comment (ankitkumarr) [inline azure/functions/_http_wsgi.py]: remove comment

**Comment by alfogo** [comment]
`import logging

import azure.functions as func
from ..FlaskApp.wsgi import application

def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
    return func.WsgiMiddleware(application).handle(req, context)`

Above code returns Exception: ModuleNotFoundError: No module named '__app__.FlaskApp' either locally on Azure

Am I missing something else?

**Comment by alfogo** [comment]
Hi,
I manage to get this working but am getting 404 when calling a test Flask app (which is uploaded  pypi)

Azure Function:
test1

Flask app:
image

Call:
image

Do I need to change something else in the configuration?

Thx

**Comment by b0lle** [comment]
Hi guys,

thank you for this peace of code. I am using this to host my flask api in azure functions.
As I recently found out there seems to be an issue with the middleware in connection with KeyVault. In my opinion the HttpLoggingPolicy class in azu.core.pipeline.policies._universal uses the logging module to log the status of the http response. Unfortunately this populates the _wsgi_error_buffer in WsgiMiddleware (C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8\WINDOWS\X64\azure\functions\_http_wsgi.py).
When this buffer is filled with any content, then the handle_error function of the WsgiMiddleware raises an error with the content of the buffer.
I do not know why the buffer was filled. But it hat to be connected because disabling the logger of the HttpLoggingPolicy (logging.getLogger('azure.core.pipeline.policies.http_logging_policy').setLevel(logging.ERROR)) solved the issue in local debugging. Unfortunately this didnt solve the issue in the deployed function.

Here are some details about the api endpoints, the access to the keyvault and the error message:

<I was unable to format the code in pthon style here. please see the attached files>
[app.py.txt](https://github.com/Azure/azure-functions-python-library/files/5964118/app.py.txt)
[az_func__init__.py.txt](https://github.com/Azure/azure-functions-python-library/files/5964119/az_func__init__.py.txt)

This is the error message (to get the error message I printed the exception in _http_wsgi.py - handle_error):
**[2021-02-11T09:19:05.653] error_message: [2021-02-11 10:18:58,855] INFO in _universal: Response status: 401**
[2021-02-11T09:19:05.739] Executed 'Functions.AdApi' (Failed, Id=0d594991-4412-4cef-9b1e-14f1d6f27602, Duration=13984ms)
[2021-02-11T09:19:05.742] System.Private.CoreLib: Exception while executing function: Functions.AdApi. System.Private.CoreLib: Result: Failure
**Exception: Exception: [2021-02-11 10:18:58,855] INFO in _universal: Response status: 401**

Stack:   File "C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 343, in _handle__invocation_request
    call_result = await self._loop.run_in_executor(
  File "C:\MyPrograms\Anaconda3\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8/WINDOWS/X64\azure_functions_worker\dispatcher.py", line 480, in __run_sync_func
    return func(**params)
  File "C:\Azure\dea-anomalydetection-prod-func\AdApi\__init__.py", line 19, in

 main
    raise ex
  File "C:\Azure\dea-anomalydetection-prod-func\AdApi\__init__.py", line 16, in main
    return func.WsgiMiddleware(application).handle(req, context)
  File "C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8/WINDOWS/X64\azure\functions\_http_wsgi.py", line 161, in handle
    return self._handle(req, context)
  File "C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8/WINDOWS/X64\azure\functions\_http_wsgi.py", line 174, in _handle
    self._handle_errors()
  File "C:\Program Files\Microsoft\Azure Functions Core Tools\workers\python\3.8/WINDOWS/X64\azure\functions\_http_wsgi.py", line 184, in _handle_errors
    raise Exception(error_message)

As you can see in the exception: 
The content seems to be from "INFO in _universal:" So I disabled the logging of the function:
`logger = logging.getLogger('azure.core.pipeline.policies.http_logging_policy')`
`logger.setLevel(logging.ERROR)`
which solved the issue local (VS Studio Code) but not deployed

**Comment by elasticdotventures** [comment]
> The following code should do the trick.
> 
> ```python
> import logging
> 
> import azure.functions as func
> from ..Django.wsgi import DjangoApplication
> 
> main = func.WsgiMiddleware(DjangoApplication).main
> ```

Trying to wrap my head around Azure Functions & Django and WSGI at the same time and this helpful snippet makes little/no sense to me.   (Perhaps I'm being too literal) for example @Hazhzeng do you mean to use "..Django.wsgi" or is that just a lazy notation?  I'm not sure what the ".." operator does in Python, so I'm going to assume it's a typo/invalid and it's not equivalent to the javascript spread operator, maybe your python-fu is stronger than mine. 

I'm going to assume that code snippet belongs in __init__.py  (or whatever file is referenced by the function.json), and that it's bindings should be of "type": "httpTrigger" and probably "authLevel":"anonymous"

because manage.py is the command line utility for admin tasks

In reality I think it should probably look something more like: 
```python

from django.core.wsgi import get_wsgi_application
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'the.settings')
application = get_wsgi_application()
main = func.WsgiMiddleware(application).main

```
^^^ note: YMMV .. I haven't actually gotten this working YET (so the lines above may be wrong), but I'm pretty confident the earlier sample above doesn't work either. 

I'm also going to assume that on Azure functions the host.json needs to have the routePrefix disabled such as:
    "extensions": {
        "http": {
            "routePrefix": ""
        }
    }
as described in the  "Customize the http endpoint" section of the Azure Functions reference
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=python#customize-the-http-endpoint

But again, I can't figure out how to get Django to load from an Azure function but I've put in enough time so I'm giving up and heading back to a Docker instance.

**Comment by DharakOne** [comment]
How do create wsgi object  in flask?

Could someone to give me example?

from ..FlaskApp.wsgi import application


```python
import logging

import azure.functions as func
from ..FlaskApp.wsgi import application

```

**Comment by stefanushinardi** [comment]
@DharakOne this example might clear out some things: https://github.com/Hazhzeng/functions-wsgi-demo. Let me know if it is not clear

**Comment by echennells** [comment]
> @DharakOne this example might clear out some things: https://github.com/Hazhzeng/functions-wsgi-demo. Let me know if it is not clear

Isn't the example using the azf-wsgi library and the old style?

https://github.com/Hazhzeng/functions-wsgi-demo/blob/master/HttpTrigger/__init__.py

**Comment by eelwk** [comment]
For anyone wishing for a very simple starting example implementation, try this:

`app.py` file (this is your simple hello world flask app)
path to file ..function-flask/api/app.py
```from flask import Flask
import json

application = Flask(__name__)

@application.route("/api/httpTrigger")
def home():
    return "Hello, from Flask!"
```
I used an http Trigger function, cleverly named httpTrigger, thus the routing above.
My function app's `__init.py__`
path to file ..function-flask/httpTrigger/__init__.py
```
import logging

import azure.functions as func
from api.app import application

def main(req: func.HttpRequest, context: func.Context) -> func.HttpResponse:
    return func.WsgiMiddleware(application).handle(req, context)
```

Running the function app locally using the function runtime core tools, if I hit this URL:
http://localhost:7071/api/httpTrigger

The response I get back is a 200 OK and "Hello, from Flask!"

**Comment by IvanFerrante92** [comment]
Hi guys! 👋
For those looking for an easy implementation example with `flask_restx` and `API` registration: https://github.com/IvanFerrante92/azfunc-wsgi-example/tree/main/Azfn-Wsgi/AzfnWSGI

**Comment by wanghaisheng** [comment]
@eelwk 
hey there  i got my app.py like this, how to adjust this
```


import json

application = Flask(__name__)

def home():
    return "Hello, from Flask!"


if __name__ == "__main__":

    app.add_url_rule('/', 'webio_view', webio_view(partial(home, lang='')),
                     methods=['GET', 'POST', 'OPTIONS'])
   app.run()
```

**Pull request merged**